tsl/src/remote/dist_copy.c

```diff
@@ -54,13 +45,32 @@ typedef struct CopyDimensionInfo
 	int32 atttypmod;
 } CopyDimensionInfo;
 
+typedef struct DataNodeConnection
+{
+	TSConnectionId id;
+	TSConnection *connection;
+} DataNodeConnection;
+
 /* This contains information about connections currently in use by the copy as well as how to create
  * and end the copy command.
  */
 typedef struct CopyConnectionState
 {
-	List *cached_connections;
+	/*
+	 * Cached connections to data nodes.
+	 * Why do we need another layer of caching, when there is dist_txn layer
+	 * already? The API it provides is one function that "does everything
+	 * automatically", namely it's going to stop the COPY each time we request
+	 * the connection. This is not something we want to do for each row when
+	 * we're trying to do bulk copy.
+	 */
+	List *data_node_connections;
```

Comment:
We might not need this list if we instead use the standard connection cache (see comment below).

Reply:
It's ok to not refactor to use the connection cache in this PR, although I think this is what we should aim for as to not have multiple ways of dealing with connections.

Reply:
The problem with using `remote_connection_cache` directly is that the remote chunk creation (`chunk_api_create_on_data_nodes`) uses the `dist_txn` layer.